File: backend/retrieval_graph/researcher_graph/graph.py
"""Researcher graph used in the conversational retrieval system as a subgraph.

This module defines the core structure and functionality of the researcher graph,
which is responsible for generating search queries and retrieving relevant documents.
"""

from typing import cast

from langchain_core.documents import Document
from langchain_core.runnables import RunnableConfig
from langgraph.constants import Send
from langgraph.graph import END, START, StateGraph
from typing_extensions import TypedDict

from backend import retrieval
from backend.retrieval_graph.configuration import AgentConfiguration
from backend.retrieval_graph.researcher_graph.state import QueryState, ResearcherState
from backend.utils import load_chat_model


async def generate_queries(
    state: ResearcherState, *, config: RunnableConfig
) -> dict[str, list[str]]:
    """Generate search queries based on the question (a step in the research plan).

    This function uses a language model to generate diverse search queries to help answer the question.

    Args:
        state (ResearcherState): The current state of the researcher, including the user's question.
        config (RunnableConfig): Configuration with the model used to generate queries.

    Returns:
        dict[str, list[str]]: A dictionary with a 'queries' key containing the list of generated search queries.
    """

    class Response(TypedDict):
        queries: list[str]

    configuration = AgentConfiguration.from_runnable_config(config)
    structured_output_kwargs = (
        {"method": "function_calling"} if "openai" in configuration.query_model else {}
    )
    model = load_chat_model(configuration.query_model).with_structured_output(
        Response, **structured_output_kwargs
    )
    messages = [
        {"role": "system", "content": configuration.generate_queries_system_prompt},
        {"role": "human", "content": state.question},
    ]
    response = cast(
        Response, await model.ainvoke(messages, {"tags": ["langsmith:nostream"]})
    )
    return {"queries": response["queries"]}


async def retrieve_documents(
    state: QueryState, *, config: RunnableConfig
) -> dict[str, list[Document]]:
    """Retrieve documents based on a given query.

    This function uses a retriever to fetch relevant documents for a given query.

    Args:
        state (QueryState): The current state containing the query string.
        config (RunnableConfig): Configuration with the retriever used to fetch documents.

    Returns:
        dict[str, list[Document]]: A dictionary with a 'documents' key containing the list of retrieved documents.
    """
    with retrieval.make_retriever(config) as retriever:
        response = await retriever.ainvoke(state.query, config)
        return {"documents": response}


def retrieve_in_parallel(state: ResearcherState) -> list[Send]:
    """Create parallel retrieval tasks for each generated query.

    This function prepares parallel document retrieval tasks for each query in the researcher's state.

    Args:
        state (ResearcherState): The current state of the researcher, including the generated queries.

    Returns:
        Literal["retrieve_documents"]: A list of Send objects, each representing a document retrieval task.

    Behavior:
        - Creates a Send object for each query in the state.
        - Each Send object targets the "retrieve_documents" node with the corresponding query.
    """
    return [
        Send("retrieve_documents", QueryState(query=query)) for query in state.queries
    ]


# Define the graph
builder = StateGraph(ResearcherState)
builder.add_node(generate_queries)
builder.add_node(retrieve_documents)
builder.add_edge(START, "generate_queries")
builder.add_conditional_edges(
    "generate_queries",
    retrieve_in_parallel,  # type: ignore
    path_map=["retrieve_documents"],
)
builder.add_edge("retrieve_documents", END)
# Compile into a graph object that you can invoke and deploy.
graph = builder.compile()
graph.name = "ResearcherGraph"

File: backend-js/src/retrieval_graph/researcher_graph/graph.ts
/**
 * Researcher graph used in the conversational retrieval system as a subgraph.
 *
 * This module defines the core structure and functionality of the researcher graph,
 * which is responsible for generating search queries and retrieving relevant documents.
 */

import { RunnableConfig } from '@langchain/core/runnables'
import { StateGraph, START, END, Send } from '@langchain/langgraph'
import { z } from 'zod'
import {
  ResearcherStateAnnotation,
  QueryStateAnnotation,
  QueryState,
} from './state.js'
import { getAgentConfiguration } from '../configuration.js'
import { getGenerateQueriesSystemPrompt } from '../prompts.js'
import { loadChatModel } from '../../utils.js'
import { makeRetriever } from '../../retrieval.js'
/**
 * Schema for query generation response
 */
const GenerateQueriesSchema = z.object({
  queries: z.array(z.string()).describe('List of search queries to execute'),
})

/**
 * Generate search queries based on the question (a step in the research plan).
 *
 * This function uses a language model to generate diverse search queries to help answer the question.
 *
 * @param state - The current state of the researcher, including the user's question
 * @param config - Configuration with the model used to generate queries
 * @returns Updated state with generated queries
 */
async function generateQueries(
  state: typeof ResearcherStateAnnotation.State,
  config?: RunnableConfig,
): Promise<Partial<typeof ResearcherStateAnnotation.State>> {
  const configuration = getAgentConfiguration(config)
  const systemPrompt = await getGenerateQueriesSystemPrompt()

  // Determine if we should use function calling method
  const useFunctionCalling = configuration.queryModel.includes('openai')

  const model = loadChatModel(configuration.queryModel)
  const structuredModel = model.withStructuredOutput(GenerateQueriesSchema, {
    method: useFunctionCalling ? 'functionCalling' : 'json_schema',
  })

  const messages = [
    { role: 'system' as const, content: systemPrompt },
    { role: 'human' as const, content: state.question },
  ]

  const response = await structuredModel.invoke(messages, {
    ...config,
    tags: ['langsmith:nostream'],
  })

  return {
    queries: response.queries,
  }
}

/**
 * Retrieve documents based on a given query.
 *
 * This function uses a retriever to fetch relevant documents for a given query.
 *
 * @param state - The current state containing the query string
 * @param config - Configuration with the retriever used to fetch documents
 * @returns Updated state with retrieved documents
 */
async function retrieveDocuments(
  state: typeof QueryStateAnnotation.State,
  config?: RunnableConfig,
): Promise<Partial<typeof QueryStateAnnotation.State>> {
  const retriever = await makeRetriever(config)

  const documents = await retriever.invoke(state.query, config)

  return {
    documents,
    queryIndex: state.queryIndex,
  }
}

/**
 * Create parallel retrieval tasks for each generated query.
 *
 * This function prepares parallel document retrieval tasks for each query in the researcher's state.
 *
 * @param state - The current state of the researcher, including the generated queries
 * @returns List of Send objects, each representing a document retrieval task
 */
function retrieveInParallel(
  state: typeof ResearcherStateAnnotation.State,
): Send<string, QueryState>[] {
  return state.queries.map(
    (query, index) =>
      new Send<string, QueryState>('retrieve_documents', {
        query,
        queryIndex: index,
        documents: [],
      } as QueryState),
  )
}

/**
 * Build and compile the researcher graph
 */
const builder = new StateGraph(ResearcherStateAnnotation)
  .addNode('generate_queries', generateQueries)
  .addNode('retrieve_documents', retrieveDocuments)
  .addEdge(START, 'generate_queries')
  .addConditionalEdges('generate_queries', retrieveInParallel, [
    'retrieve_documents',
  ])
  .addEdge('retrieve_documents', END)

export const graph = builder.compile()
graph.name = 'ResearcherGraph'
